### ISSUE TITLE
ONNX gpu io binding buffers corrupted

### ISSUE THREAD

**Issue by viboga** [microsoft/onnxruntime#10103]
**Describe the bug**
Onnx inference on GPU provides performance boost with io binding. The buffers for outputs of a model are pre-allocated so they can be used over multiple iterations. This buffer when used repeatedly with varying shapes causes corruption of the order or corruption of memory itself. The buffers are created on GPU with torch.zeros().

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows Server 2016 DataCenter
- ONNX Runtime installed from (source or binary): source 
- ONNX Runtime version: 1.10 (latest as of 12/15/2021)
- Python version: python 8
- Visual Studio version (if applicable): VS 2019
- GCC/Compiler version (if compiling from source): build from source, VS compiler
- CUDA/cuDNN version: 11.3
- GPU model and memory: V100, 32GB

**To Reproduce**
- Describe steps/code to reproduce the behavior:
In order to fetch multiple outputs from multiple inferences, have a pre-allocated buffer of larger size than any of the intended outputs. 
1. For each iteration
Fetch the outputs from buffer using reshape logic as follows:
buffer.view(-1)[:target_size].view(target_shape)
2. If the size of target changes from iteration to iteration the expected output is different from when inference run without io binding after couple of iterations.

- Attach the ONNX model to the issue (where applicable) to expedite investigation
Model issue was reproduced on modified GPT2 with BxSx50297 logits size.

**Expected behavior**
The outputs are same with and without io binding.

**Screenshots**

**Additional context**:
I am describing the exact steps I used for a modified GPT2:
logits size of Batch_size x Sequence_length x 50297 (vocab) for input size of Batch_size x Sequence_length. There are other inputs attention_mask(bxs) (lower triangle of 1's)
position_ids(bxs) (ascending index from 0 - [0,1,2,..s-1]
and past_state(2xbx32xsx128) - first iteration NA, from second iterations outputs from previous run.

1. Allocate buffer for logits with size 4 (batch_size) x 1024 (sequence_length) x 50297.
2. For the first iteration let b = 4 and s = 10 (4 batches are replicas of just one). After inference fetch the logits using
target_size = [4,10,50297]
result = logits_buffer.view(-1)[:np.prod(target_size)].view(target_size)

3. For the second iteration let the batch_size be 4, sequence_length be 1. Same as step 2, its a replica
Fetch the logits using 
target_size = [4, 1, 50297]
result = logits_buffer.view(-1)[:np.prod(target_size)].view(target_size)

Expected output is all four batches are similar in second iteration: [0,1,50297] == [1,1,50297] == [2,1,50297] == [3,1,50297]
but this is not the case.

**Comment by viboga** [comment]
Clone with buffer when reshaped - this would resolve the issue with no significant increase in latency:

result = logits_buffer.view(-1)[:np.prod(target_size)].view(target_size).clone()
